enzy_htp/structure/atom.py

```diff
@@ -100,23 +109,136 @@ def b_factor(self, val):
+                    cnt_atomnames = chem.residue.RESIDUE_CONNECTIVITY_MAP_CTERMINAL[parent_residue.name][self.name]
+                else:
+                    cnt_atomnames = chem.residue.RESIDUE_CONNECTIVITY_MAP[parent_residue.name][self.name]
+        else:
+            _LOGGER.error(f"getting connectivity of non-canonical residue {self.parent}")
+            sys.exit(1)
+        for name in cnt_atomnames:
+            try:
+                if name not in ["-1C", "+1N"]:
+                    cnt_atom = parent_residue.find_atom_name(name)
+                if name == "-1C":
+                    cnt_resi = parent_residue.chain.find_residue_idx(parent_residue.idx-1)
+                    cnt_atom = cnt_resi.find_atom_name("C")
+                if name == "+1N":
+                    cnt_resi = parent_residue.chain.find_residue_idx(parent_residue.idx+1)
+                    cnt_atom = cnt_resi.find_atom_name("N")
+                connect.append(cnt_atom)
+            except ResidueDontHaveAtom as e:
+                _LOGGER.warning(f"missing connecting atom {e.atom_name} of {self}. Structure maybe incomplete.")
+        self._connect = connect
+        return self._connect
+    #endregion
+
+    #region === Getter-Property (ref) ===
+    def radius(self, method: str = "ionic") -> float:
+        """Gets the atomic radii with specified definition.
+        Args:
+            method: the method to determine radius the atom.
+            (current available keywords)
+                ionic: (ionic radius) for both metal and donor atom
+                vdw: (Van Der Waals radius) for both metal and donor atom
+        Returns:
+            a float of the radius
+        """
+        radius = chem.get_atom_radii(self.element, method)
+        return radius
+
+    @dispatch
+    def distance_to(self, point: Atom) -> float:
+        """Get the distance to the other atom or a point."""
+        return mh.get_distance(self.coord, point.coord)
+
+    @dispatch
+    def distance_to(self, point: tuple) -> float: # pylint: disable=function-redefined
+        """Get the distance to the other atom or a point."""
+        return mh.get_distance(self.coord, point)
+
+    def attached_protons(self) -> List[Atom]:
+        """find all protons attached to self"""
+        result = list(filter(lambda a: a.element == "H", self.connect))
+        return result
+
+    #endregion
+
+    #region === Check ===
+    def is_donor_atom(self) -> bool:
+        """check if the atom is a donor atom to a coordination center"""
+        return self.name in chem.metal.DONOR_ATOM_LIST
+    
+    def is_connected(self) -> bool:
+        """check if self is in the connected state"""
+        return self._connect is not None
     #endregion
 
     #region == Special ==
     def __str__(self):
         return f"Atom({self._name}, {self._idx}, {self._coord}, {self._parent}, {self._b_factor}, {self._element}, {self._charge} )"
     #endregion
 
+    @dispatch
+    def _(self):
```

Comment:
What is this?

Reply:
This is required by plum-dispatch when use it in a class